src/silo/query_engine/actions/fasta.cpp

```diff
@@ -162,87 +167,148 @@ void Fasta::addSequencesToResultsForPartition(
    addSequencesFromResultTableToJson(
-      results, connection, result_table_name, sequence_names, database_partition, number_of_values
+      results, connection, result_table_name, sequence_names, database_partition, num_result_rows
    );
 
-   (void)query(connection, fmt::format("DROP TABLE {};", result_table_name));
-   (void)query(connection, fmt::format("DROP TABLE {};", key_table_name));
+   // XX don't we need guaranteed cleanup for those?
+   query(connection, fmt::format("DROP TABLE {};", result_table_name));
+   query(connection, fmt::format("DROP TABLE {};", key_table_name));
+
+   return last_row_id;
 }
 
+}  // namespace
+
+// Note: fasta_aligned.cpp has its own PARTITION_CHUNK_SIZE
+static const size_t PARTITION_CHUNK_SIZE = 10000;
+
 QueryResult Fasta::execute(const Database& database, std::vector<OperatorResult> bitmap_filter)
    const {
    for (const std::string& sequence_name : sequence_names) {
       CHECK_SILO_QUERY(
          database.unaligned_nuc_sequences.contains(sequence_name),
          "Database does not contain an unaligned sequence with name: '" + sequence_name + "'"
-      )
-   }
-
-   const std::string& primary_key_column = database.database_config.schema.primary_key;
-
-   size_t total_count = 0;
-   for (auto& filter : bitmap_filter) {
-      total_count += filter->cardinality();
+      );
    }
-   CHECK_SILO_QUERY(
-      total_count <= SEQUENCE_LIMIT,
-      fmt::format("Fasta action currently limited to {} sequences", SEQUENCE_LIMIT)
-   );
 
-   QueryResult results;
-   results.query_result.reserve(total_count);
-
-   for (uint32_t partition_index = 0; partition_index < database.partitions.size();
-        ++partition_index) {
-      const auto& database_partition = database.partitions[partition_index];
-      const auto& bitmap = bitmap_filter[partition_index];
-
-      addSequencesToResultsForPartition(results, database_partition, bitmap, primary_key_column);
-   }
+   size_t current_partition = 0;
+   // The unprocessed part of the result row numbers, counting from 0
+   // from the start of the current partition:
+   std::optional<Range<uint32_t>> remaining_result_row_indices;
+   return QueryResult{[sequence_names = sequence_names,
+                       remaining_result_row_indices,
+                       bitmap_filter =
+                          std::make_shared<std::vector<OperatorResult>>(std::move(bitmap_filter)),
+                       &database,
+                       current_partition](std::vector<QueryResultEntry>& results) mutable {
+      while (current_partition < database.partitions.size()) {
+         // We drain bitmap_filter as we process the query (because
+         // the only way to get an iterator over a bitmap is from the
+         // beginning?)! And `remaining_result_row_indices` is really
+         // only for show now.
+         auto& bitmap = (*bitmap_filter)[current_partition];
+         if (!remaining_result_row_indices) {
```

Comment:
Possibly, I'm not 100% happy with that code either, but there's still other work to be done for the streaming, can I write a TODO comment saying so (and connected issue if you like and get back to it later)?

Reply:
My comment referred to simply spelling out the `!` operator on the concrete type, since it's `std::optional<Range<uint32_t>>` and not a boolean.

I know what `!` does on a boolean, but I would need to look it up for this type. Is it equivalent to `!remaining_result_row_indices.has_value()`?

Reply:
Yes, optional is overloaded to work as boolean [1]; IIRC there are various cases of that in the code base already.

[1] https://en.cppreference.com/w/cpp/utility/optional/operator_bool

Reply:
I guess you mean to say that you don't like that overload and prefer `has_value()` to get an actual boolean. Using integers and pointers in boolean context is very normal in C and C++, too, though, as is truthy in Python etc., but I'm also sympathetic to precise type matching and will use `has_value()`.